src/Collections/Sequences/Actions.dfy

```diff
@@ -0,0 +1,261 @@
+    // Idealy, it would be great to remove this
+    // and simply prove everything about `res`.
+    // However in practice this has proven to be difficult.
+    // Given how flexible FlatMap is,
+    // there may not be a prcatical general solution.
+    returns (res: Result<seq<R>, E>, ghost parts: seq<seq<R>>)
+    requires forall i | i in s :: action.Requires(i)
+    ensures
+      res.Success?
+    ==>
+      && |s| == |parts|
+      && res.value == Flatten(parts)
+      && (forall i :: 0 <= i < |s|
+      ==>
+        && action.Ensures(s[i], Success(parts[i]))
+        && multiset(parts[i]) <= multiset(res.value)
+      )
+  {
+    parts := [];
+    var rs := [];
+    for i := 0 to |s|
+      invariant |s[..i]| == |parts|
+      invariant forall j ::
+        && 0 <= j < i
+      ==>
+        && action.Ensures(s[j], Success(parts[j]))
+        && multiset(parts[j]) <= multiset(rs)
+      invariant Flatten(parts) == rs
+    {
+      var maybeR := action.Invoke(s[i]);
+      if maybeR.Failure? {
+        return Failure(maybeR.error), parts;
+      }
+      var r := maybeR.value;
+      rs := rs + r;
+      LemmaFlattenConcat(parts, [r]);
+      parts := parts + [r];
+    }
+    return Success(rs), parts;
+  }
+
+  method Filter<A>(
+    action: Action<A, bool>,
+    s: seq<A>
+  )
+    returns (res: seq<A>)
+    requires forall i | i in s :: action.Requires(i)
+    ensures |s| >= |res|
+    ensures
+      forall j ::
+        j in res
+      ==>
+        && j in s
+        && action.Ensures(j, true)
+  {
+    var rs := [];
+    for i := 0 to |s|
+      invariant |s[..i]| >= |rs|
+      invariant forall j ::
+        j in rs
+      ==>
+        && j in s
+        && action.Ensures(j, true)
+    {
+      var r := action.Invoke(s[i]);
+      if r {
+        rs := rs + [s[i]];
+      }
+    }
+    return rs;
+  }
+
+  method FilterWithResult<A, E>(
+    action: Action<A, Result<bool, E>>,
+    s: seq<A>
+  )
+    returns (res: Result<seq<A>, E>)
+    requires forall i | i in s :: action.Requires(i)
+    ensures
+      res.Success?
+    ==>
+      && |s| >= |res.value|
+      && forall j ::
+        j in res.value
+      ==>
+        && j in s
+        && action.Ensures(j, Success(true))
+  {
+    var rs := [];
+    for i := 0 to |s|
+      invariant |s[..i]| >= |rs|
+      invariant forall j ::
+        j in rs
+      ==>
+        && j in s
+        && action.Ensures(j, Success(true))
+    {
+      var maybeR := action.Invoke(s[i]);
+      if maybeR.Failure? {
+        return Failure(maybeR.error);
+      }
+      var r := maybeR.value;
+      if r {
+        rs := rs + [s[i]];
+      }
+    }
+    return Success(rs);
+  }
+
+  method ReduceToSuccess<A, B, E>(
+    action: Action<A, Result<B, E>>,
+    s: seq<A>
+  )
+    returns (res: Result<B, seq<E>>)
+    requires forall i | i in s :: action.Requires(i)
+    ensures
+      res.Success?
+    ==>
+      exists a | a in s :: action.Ensures(a, Success(res.value))
```

Comment:
Don't you want to make this stronger and say it's the FIRST such element?

Reply:
Yes, I do! But I did not know how.

Reply:
`exists i | 0 <= i < |s| :: action.Ensures(s[i], Success(res.value)) && forall j | 0 <= j < i ==> exist f | action.Ensures(s[i], Failure(f)) `
 
Maybe there is a better way?